src/moves/MultiParticle.h

```diff
@@ -457,64 +460,95 @@ inline void MultiParticle::Accept(const uint rejectState, const ulong step)
   GOMC_EVENT_STOP(1, GomcProfileEvent::ACC_MULTIPARTICLE);
 }
 
-inline XYZ MultiParticle::CalcRandomTransform(XYZ const &lb, double const max, uint molIndex)
+inline XYZ MultiParticle::CalcRandomTransform(bool &forceInRange, XYZ const &lb,
+                                              double const max, uint molIndex)
 {
   XYZ lbmax = lb * max;
-  XYZ num, randnums;
-  randnums = r123wrapper.GetRandomCoords(molIndex);
-  if(std::abs(lbmax.x) > MIN_FORCE && std::abs(lbmax.x) < MAX_FORCE) {
-    num.x = log(exp(-1.0 * lbmax.x) + 2 * randnums.x * sinh(lbmax.x)) / lb.x;
-  } else {
-    double rr = randnums.x * 2.0 - 1.0;
-    num.x = max * rr;
-  }
-
-  if(std::abs(lbmax.y) > MIN_FORCE && std::abs(lbmax.y) < MAX_FORCE) {
-    num.y = log(exp(-1.0 * lbmax.y) + 2 * randnums.y * sinh(lbmax.y)) / lb.y;
-  } else {
-    double rr = randnums.y * 2.0 - 1.0;
-    num.y = max * rr;
-  }
-
-  if(std::abs(lbmax.z) > MIN_FORCE && std::abs(lbmax.z) < MAX_FORCE) {
-    num.z = log(exp(-1.0 * lbmax.z) + 2 * randnums.z * sinh(lbmax.z)) / lb.z;
-  } else {
-    double rr = randnums.z * 2.0 - 1.0;
-    num.z = max * rr;
+  // XYZ default constructor initializes to (0.0, 0.0, 0.0)
+  XYZ val;
+  forceInRange = std::abs(lbmax.x) > MIN_FORCE && std::abs(lbmax.x) < MAX_FORCE &&
+                 std::abs(lbmax.y) > MIN_FORCE && std::abs(lbmax.y) < MAX_FORCE &&
+                 std::abs(lbmax.z) > MIN_FORCE && std::abs(lbmax.z) < MAX_FORCE;
+
+  if (forceInRange) {
+    XYZ randnums;
+#ifdef _OPENMP
+    //Even though we call different random123 functions, they all change c[0]
+    //and use the same rng() function, so they all need to be in the same
+    //critical section (not different names).
+    #pragma omp critical
+#endif
```

Comment:
As discussed, in later version we can implement the random123 CPU version, similar to GPU version to avoid using `omp critical` overhead.

Reply:
Thanks. Yes, I will test and verify that change before merging the patch. I'll push an update if I can get it working without a critical section.